web/packages/teleport/src/lib/useMfa.ts

```diff
@@ -16,220 +16,170 @@
+  parseMfaChallengeJson,
+} from 'teleport/services/auth';
 import auth from 'teleport/services/auth/auth';
+import {
+  DeviceType,
+  getMfaChallengeOptions,
+  MfaAuthenticateChallenge,
+  MfaChallengeResponse,
+  MfaOption,
+} from 'teleport/services/mfa';
+
+export type MfaProps = {
+  req?: CreateAuthenticateChallengeRequest;
+  isMfaRequired?: boolean | null;
+};
 
-export function useMfa(emitterSender: EventEmitterMfaSender): MfaState {
-  const [state, setState] = useState<{
-    errorText: string;
-    addMfaToScpUrls: boolean;
-    webauthnPublicKey: PublicKeyCredentialRequestOptions;
-    ssoChallenge: SSOChallenge;
-    totpChallenge: boolean;
-  }>({
-    addMfaToScpUrls: false,
-    errorText: '',
-    webauthnPublicKey: null,
-    ssoChallenge: null,
-    totpChallenge: false,
-  });
-
-  function clearChallenges() {
-    setState(prevState => ({
-      ...prevState,
-      totpChallenge: false,
-      webauthnPublicKey: null,
-      ssoChallenge: null,
-    }));
-  }
-
-  function onSsoAuthenticate() {
-    if (!state.ssoChallenge) {
-      setState(prevState => ({
-        ...prevState,
-        errorText: 'Invalid or missing SSO challenge',
-      }));
-      return;
-    }
-
-    // try to center the screen
-    const width = 1045;
-    const height = 550;
-    const left = (screen.width - width) / 2;
-    const top = (screen.height - height) / 2;
-
-    // these params will open a tiny window.
-    const params = `width=${width},height=${height},left=${left},top=${top}`;
-    window.open(state.ssoChallenge.redirectUrl, '_blank', params);
-  }
-
-  function onWebauthnAuthenticate() {
-    if (!window.PublicKeyCredential) {
-      const errorText =
-        'This browser does not support WebAuthn required for hardware tokens, \
-      please try the latest version of Chrome, Firefox or Safari.';
-
-      setState({
-        ...state,
-        errorText,
-      });
-      return;
-    }
-
-    auth
-      .getMfaChallengeResponse({
-        webauthnPublicKey: state.webauthnPublicKey,
-      })
-      .then(res => {
-        setState(prevState => ({
-          ...prevState,
-          errorText: '',
-          webauthnPublicKey: null,
-        }));
-        emitterSender.sendWebAuthn(res.webauthn_response);
-      })
-      .catch((err: Error) => {
-        setErrorText(err.message);
-      });
-  }
-
-  const waitForSsoChallengeResponse = useCallback(
-    async (
-      ssoChallenge: SSOChallenge,
-      abortSignal: AbortSignal
-    ): Promise<void> => {
-      const channel = new BroadcastChannel(ssoChallenge.channelId);
+// useMfa prepares an MFA state designed for use with AuthnDialog.
+// 1. Call getMfaChallengeResponse() where MFA may be required.
+//    If MFA is not required, it should be a noop, returning null.
+// 2. Setup AuthnDialog using the MFA state, to show a dialog when
+//    mfaAttempt.state === 'processing'.
+export function useMfa({ req, isMfaRequired }: MfaProps): MfaState {
```

Comment:
Hmm, is this hook actually tested anywhere? If it's not, adding a unit test to it would be a good idea, since there's a lot of non-trivial logic here.

Reply:
235d2e118a793564a28cd7bf4aa849260ef97590